File: backend/app/schemas/driver.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, field_validator


class DriverBase(BaseModel):
    full_name: str
    employee_code: str | None = None
    national_id: str
    phone: str
    email: str | None = None
    vehicle_type: str
    vehicle_plate: str
    license_number: str
    license_expiry: datetime | None = None
    status: str = "Active"
    availability: str = "Available"
    branch_id: int | None = None
    is_active: bool = True

    @field_validator("license_expiry", mode="before")
    @classmethod
    def normalize_license_expiry(cls, value):
        if value is None:
            return None
        if isinstance(value, str):
            value = value.strip()
            if value == "":
                return None
        return value


class DriverCreate(DriverBase):
    pass


class DriverUpdate(DriverBase):
    pass


class DriverResponse(DriverBase):
    company_id: int
    branch_name: str | None = None
    assigned_shipments_count: int = 0
    delivered_today_count: int = 0
    pending_deliveries_count: int = 0
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    updated_at: datetime

File: frontend/types/driver.ts
export type DriverCreatePayload = {
  full_name: string
  employee_code?: string
  national_id: string
  phone: string
  email?: string
  vehicle_type: string
  vehicle_plate: string
  license_number: string
  license_expiry?: string
  status: string
  availability: string
  branch_id?: number | null
  is_active: boolean
}

export type DriverResponse = DriverCreatePayload & {
  id: number
  company_id: number
  branch_name?: string | null
  assigned_shipments_count: number
  delivered_today_count: number
  pending_deliveries_count: number
  created_at: string
  updated_at: string
}

export type DriverListItem = DriverResponse
